### PR TITLE
Macro for creating enum classes

### PR DESCRIPTION
**Pull request by johnomotani**
After the discussion on #1835 I was thinking about using an `enum class` as an option. The way we are using `enum class`es at the moment has some boiler-plate to define `toString` and `enumFromString` functions. Using some magic copied from `bout/macro_for_each.hxx`, it seems to be possible to replace that with a macro so you can write something like
```
BOUT_ENUM_CLASS(myenum, foo, bar);
```
and get an `enum class` with possible values `myenum::foo` and `myenum::bar`, along with `toString(myenum e)` and `myenumFromString(std::string s)` functions. I've also put in overloads for `Options::as<>()` and `ostream& operator<<` so that the `enum class` can be loaded from options, e.g.
```
myenum e = options["myenum_option"].withDefault(myenum::foo);
```

I'd also like to be able to assign a string to the `enum class`, but it doesn't seem to be possible to override `operator=` for an `enum class`.

Before I think about this any more: is this a useful feature, or an abuse of macros?

### CODE REVIEW

**Comment by ZedThree** [comment]
I think this is probably essential! Worthwhile looking over some other implementations of compiletime enum reflection:

- https://github.com/aantron/better-enums
- https://github.com/Neargye/magic_enum
- https://github.com/quicknir/wise_enum
- https://github.com/therocode/meta_enum
- https://github.com/BlackMATov/enum.hpp

Some of these are C++17 only, but could be backported depending on what they use; some make an actual `class` rather than an `enum`, which might be painful for the derivative template machinery.
We might be able to use one of those wholesale, or else adapt to our needs.

I've almost finished adding `fmt` to BOUT++, which could also be useful here. It has `fmt::basic_string_view` which is an implementation of C++17's `std::string_view` that some of the above implementations use.

**Comment by bendudson** [comment]
Thanks @johnomotani I think this is both a wonderful abuse of macros, and a very useful feature. The overload of `Options::as` is very nice.

**Comment by d7919** [comment]
Can the work in progress label be removed?

**Comment by johnomotani** [comment]
Documentation of the new macro is definitely still needed.

@ZedThree - are there any useful changes to make right now from your earlier comment, or do you think it is easy enough to upgrade later?

File: include/bout/bout_enum_class.hxx [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    _BOUT_ENUM_CLASS_MAP_ARGS(_ENUM_CLASS_STR, enumname, __VA_ARGS__)      \
+  };                                                                       \
+  auto found = toString_map.find(e);                                       \
+  if (found == toString_map.end()) {                                       \
+    throw BoutException("Did not find enum %d", static_cast<int>(e));      \
+  }                                                                        \
+  return found->second;                                                    \
+}                                                                          \
+                                                                           \
+inline enumname _MAKE_FROMSTRING_NAME(enumname)(std::string s) {           \
```
Comment (ZedThree) [inline include/bout/bout_enum_class.hxx]: `std::string s` -> `const std::string& s`

**Comment by ZedThree** [comment]
Adding the fmt stuff can be done later once I work out how to do it.

One thing to change now though is that identifiers beginning with an underscore followed by a capital are reserved, so maybe change those -- I suggest just prefixing everything with `BOUT_`.

**Comment by ZedThree** [comment]
Also, might need to use this trick to get `_BOUT_ENUM_CLASS_MAP_ARGS` working:
https://github.com/boutproject/BOUT-dev/pull/1874/commits/cfb7888b495aec5031ed3bbe80436c007191f782

But you'll need a Windows system to test it on, so that can wait

**Pull request merged**